### PR TITLE
make sure connectionListener is not in the middle of calling dataAvai…

### PR DESCRIPTION
**Pull request by chehefen**
…lable callback for the socketConnection once connectionListenerRemoveConnection returns

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: src/source/Ice/ConnectionListener.c [hunk 1]
```diff
@@ -116,17 +117,34 @@ STATUS connectionListenerAddConnection(PConnectionListener pConnectionListener,
 STATUS connectionListenerRemoveConnection(PConnectionListener pConnectionListener, PSocketConnection pSocketConnection)
 {
     STATUS retStatus = STATUS_SUCCESS;
+    BOOL locked = FALSE;
+    const UINT64 longTimeout = 1 * HUNDREDS_OF_NANOS_IN_AN_HOUR;
```
Comment (MushMal) [inline src/source/Ice/ConnectionListener.c:121]: Use #define in the header. Const is a C++ keyword

File: src/source/Ice/ConnectionListener.c [hunk 2]
```diff
@@ -116,17 +117,34 @@ STATUS connectionListenerAddConnection(PConnectionListener pConnectionListener,
...
 
+    MUTEX_LOCK(pConnectionListener->lock);
+    locked = TRUE;
+    /* make sure connectionListenerRemoveConnection return after connectionListenerReceiveDataRoutine has picked up
+     * the change. */
+    while(ATOMIC_LOAD_BOOL(&pConnectionListener->listenerRoutineStarted) &&
+          !ATOMIC_LOAD_BOOL(&pConnectionListener->terminate) &&
+          ATOMIC_LOAD_BOOL(&pConnectionListener->connectionListChanged)) {
+        /* use longTimeout because connectionListenerReceiveDataRoutine should wake this up soon enough */
+        CVAR_WAIT(pConnectionListener->removeConnectionComplete, pConnectionListener->lock, longTimeout);
```
Comment (MushMal) [inline src/source/Ice/ConnectionListener.c:139]: I don't understand why 1 hour? Something more sensible perhaps with a handling of a timeout error?

File: src/source/Ice/ConnectionListener.c [hunk 3]
```diff
@@ -136,6 +154,7 @@ STATUS connectionListenerRemoveAllConnection(PConnectionListener pConnectionList
     BOOL locked = FALSE;
     PDoubleListNode pCurNode = NULL;
     PSocketConnection pSocketConnection = NULL;
+    const UINT64 longTimeout = 1 * HUNDREDS_OF_NANOS_IN_AN_HOUR;
```
Comment (MushMal) [inline src/source/Ice/ConnectionListener.c:157]: same - #define it

Comment (MushMal) [reply src/source/Ice/ConnectionListener.c:157]: Also, why is this function not calling connectionListenerRemoveConnection underneath it when iterating?

File: src/source/Ice/SocketConnection.c [hunk 4]
```diff
@@ -251,9 +252,10 @@ STATUS socketConnectionReadData(PSocketConnection pSocketConnection, PBYTE pBuf,
         // Unlikely that we will get sslReadRet == 0 here because socketConnectionReadData is only called when
         // socket recevies data. If ssl handshake is not done then -1 is returned.
         if (sslReadRet <= 0) {
-            sslErrorRet = SSL_get_error(pSocketConnection->pSsl, sslReadRet);
-            if (sslErrorRet != SSL_ERROR_WANT_WRITE && sslErrorRet != SSL_ERROR_WANT_READ) {
-                DLOGV("SSL_read failed with %s. Length of data already written %u", ERR_error_string(sslErrorRet, NULL), writtenBytes);
+            while ((sslErrorRet = ERR_get_error()) != 0) {
```
Comment (MushMal) [inline src/source/Ice/SocketConnection.c:255]: Perhaps deserves a comment, not sure what this returns

File: src/source/Ice/ConnectionListener.c (hunk 2 above)
Comment (chehefen) [reply src/source/Ice/ConnectionListener.c:139]: the point is it should never timeout. the connectionListenerReceiveDataRoutine should wake it up in time.

File: src/source/Ice/ConnectionListener.c (hunk 3 above)
Comment (chehefen) [reply src/source/Ice/ConnectionListener.c:157]: then we have to wait for each connection to be closed before closing the next one.

File: src/source/Ice/SocketConnection.c (hunk 4 above)
Comment (chehefen) [reply src/source/Ice/SocketConnection.c:255]: This is just the way openssl error is printed.

File: src/source/Ice/ConnectionListener.c (hunk 2 above)
Comment (MushMal) [reply src/source/Ice/ConnectionListener.c:139]: Perhaps, but if it does? What then? The worse could happen if the code "hangs". Perhaps you need to set something small and error out on a timeout.

File: src/source/Ice/ConnectionListener.c (hunk 3 above)
Comment (MushMal) [reply src/source/Ice/ConnectionListener.c:157]: Got it

File: src/source/Ice/SocketConnection.c (hunk 4 above)
Comment (MushMal) [reply src/source/Ice/SocketConnection.c:255]: This really does look fishy. I don't see anything in the description requiring to iterate: https://www.openssl.org/docs/manmaster/man3/ERR_get_error.html

Comment (chehefen) [reply src/source/Ice/SocketConnection.c:255]: the way it mentioned in the doc doesnt give the proper error.

File: src/source/Ice/ConnectionListener.c (hunk 2 above)
Comment (chehefen) [reply src/source/Ice/ConnectionListener.c]: I kind of prefer it hangs so we can get to it right away instead of chasing red herring.

Comment (chehefen) [reply src/source/Ice/ConnectionListener.c]: reduced to 5 seconds

Comment (MushMal) [reply src/source/Ice/ConnectionListener.c]: au contraire, on a production device, chasing hangs is impossible. I think you are doing the right thing in reducing and erring out.

File: src/source/Ice/ConnectionListener.c [hunk 5]
```diff
@@ -115,24 +116,46 @@ STATUS connectionListenerAddConnection(PConnectionListener pConnectionListener,
 
 STATUS connectionListenerRemoveConnection(PConnectionListener pConnectionListener, PSocketConnection pSocketConnection)
 {
-    STATUS retStatus = STATUS_SUCCESS;
+    STATUS retStatus = STATUS_SUCCESS, cvarWaitStatus = STATUS_SUCCESS;;
```
Comment (MushMal) [inline src/source/Ice/ConnectionListener.c:119]: double ;

**Pull request merged**